### PR TITLE
Adding Support for native calls in flix.

### PR DESCRIPTION
**Pull request by FlaminGuy**
Apart from discussions in https://github.com/flix/flix/issues/410 there are couple of other issues:

1. I'm not quite sure why `def f: Bool = unsafe native field java.lang.Boolean.TRUE` fails at invoking in `Linker.scala`
2. Is there a way to call void functions like the following:
`def f: Bool =
    let x = unsafe native new java.util.BitSet();
    unsafe native method java.util.BitSet.set(x, 1);
    true`
I also tried `let y = unsafe native method java.util.BitSet.set(x, 1);` but this tries to pop an object off the stack so this fails.
3. Every field with `_` in it fails at parsing stage
4. There is no way currently to fetch fields of an object. Looking at NativeField, no object is provided so only static field codes can be emitted.

### CODE REVIEW

**Comment by magnus-madsen** [comment]
Great! I will take a look soon. Good points about the unsupported features.

File: main/src/ca/uwaterloo/flix/language/phase/Codegen.scala [hunk 1]
```diff
@@ -573,11 +574,35 @@ object Codegen {
...
+      visitor.visitFieldInsn(getCode, declaration, name, ctx.descriptor(tpe))
+
+    case Expression.NativeMethod(method, args, tpe, loc) =>
+      // Evaluate arguments left-to-right and push them onto the stack.
+      args.foreach(compileExpression(ctx, visitor, entryPoint))
+      val declaration = asm.Type.getInternalName(method.getDeclaringClass)
+      val name = method.getName
+      val descriptor = asm.Type.getMethodDescriptor(method)
+      // If the method is static, use INVOKESTATIC otherwise use INVOKEVIRTUAL
+      val invokeCode = if (Modifier.isStatic(method.getModifiers)) INVOKESTATIC else INVOKEVIRTUAL
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Codegen.scala]: Let us call this `invokeInsn`,

File: main/test/ca/uwaterloo/flix/language/feature/TestNativeCall.flix [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+@test
+def nativeCall03: Bool =
+    let x = unsafe native new java.util.BitSet();
+    assertEq!(unsafe native method java.util.BitSet.toString(x), "{}")
+
+/////////////////////////////////////////////////////////////////////////////
+// Native Static Field                                                     //
+/////////////////////////////////////////////////////////////////////////////
+@test
+def nativeCall04: Bool = assertEq!(unsafe native field java.lang.Character.BYTES, 2)
```
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/language/feature/TestNativeCall.flix]: nativeField01

File: main/test/ca/uwaterloo/flix/language/feature/TestNativeCall.flix [hunk 3]
```diff
@@ -0,0 +1,91 @@
+/////////////////////////////////////////////////////////////////////////////
+// Native Constructor                                                      //
+/////////////////////////////////////////////////////////////////////////////
+@test
+def nativeCall01: Bool = assertEq!(unsafe native new java.lang.String(), "")
```
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/language/feature/TestNativeCall.flix]: nativeNew01 or nativeConstructor01

File: main/test/ca/uwaterloo/flix/language/feature/TestNativeCall.flix [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+def nativeCall06: Bool = assertEq!(unsafe native field java.lang.Long.SIZE, 64)
+
+@test
+def nativeCall07: Bool = assertEq!(unsafe native field java.lang.Float.SIZE, 32)
+
+/////////////////////////////////////////////////////////////////////////////
+// Native Static Methods                                                   //
+/////////////////////////////////////////////////////////////////////////////
+@test
+def nativeCall08: Bool = assertEq!(unsafe native method java.lang.Integer.bitCount(2), 1)
```
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/language/feature/TestNativeCall.flix]: nativeMethod01

**Comment by magnus-madsen** [comment]
The code looks good to me, modulo a few renamings of test cases. As for your comments:

**Comment by magnus-madsen** [comment]
> I'm not quite sure why def f: Bool = unsafe native field java.lang.Boolean.TRUE fails at invoking in Linker.scala

Can you post the exception or describe how it fails? If you think the bug is outside your code, you can open a new ticket.

> Is there a way to call void functions like the following:
> def f: Bool = let x = unsafe native new java.util.BitSet(); unsafe native method java.util.BitSet.set(x, 1); true
> I also tried let y = unsafe native method java.util.BitSet.set(x, 1); but this tries to pop an object off the stack so this fails.

I don't understand. Do you mean functions that "return" `void`?

> Every field with _ in it fails at parsing stage

I created #429.

> There is no way currently to fetch fields of an object. Looking at NativeField, no object is provided so only static field codes can be emitted.

Good point.  I created #430.

**Comment by FlaminGuy** [comment]
I fixed everything in your review.
As for `def f: Bool = unsafe native field java.lang.Boolean.TRUE` I get:
`Exception in thread "main" java.lang.NoSuchFieldError: TRUE
	at Root.f(Unknown Source)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at ca.uwaterloo.flix.runtime.Linker$$anon$2.invoke(Linker.scala:63)
	at ca.uwaterloo.flix.runtime.Solver$$anonfun$4$$anonfun$apply$3.apply(Solver.scala:742)
	at ca.uwaterloo.flix.runtime.Model.getConstant(Model.scala:69)
	at ca.uwaterloo.flix.runtime.Model.getConstant(Model.scala:72)
	at ca.uwaterloo.flix.Main$$anonfun$2.apply(Main.scala:126)
	at ca.uwaterloo.flix.util.Timer.<init>(Timer.scala:28)
	at ca.uwaterloo.flix.Main$.main(Main.scala:126)
	at ca.uwaterloo.flix.Main.main(Main.scala)`
I'm not sure that it is unrelated but I think it's a case. What do you think? (I just run this function in .flix file outside of any testsuite).      
And yes, I mean functions which return void. Is there a way to call them?

**Comment by magnus-madsen** [comment]
I think it is a type/boxing issue. 

I would expect that `java.lang.Boolean.TRUE` returns an object (i.e. a boxed boolean), but probably the inferred type is primitive boolean (due to the annotation: Bool) hence it cannot find a field of type primitive boolean. You can try `--Xdebug` to see what bytecode is emitted.

I wonder what would be a good solution to this.

**Comment by magnus-madsen** [comment]
If a method has a void return  type then push the unit value on the stack.

**Pull request merged**